### ISSUE TITLE
Strange behavior when reading from a queue (looping forever)

### ISSUE THREAD

**Issue by AzSeb** [OpenHFT/Chronicle-Queue#342]
Hi,
I've recently tried to integrate Chronicle Queue into one of my projects and have encountered a strange issues whilst reading data from the end of a queue. In my testing, I discovered that I need to add a Thread.sleep or a Locksupport.parkNanos to get it to work (even if it's only 1ns delay). But, having other forms of pausing (like busy looping) do not work.

If I don't add this delay the program will work with some strange results: some time the read value is 0 (although the value written in the queue is correct -  I checked it using the tool provided) and other it will simply start looping forever never to finish reading (needing to force stop the execution).

Another strange behavior being that latency is smaller whilst using a System.out.println instead of a delay (like mentioned above) even though the delay is smaller than the time it takes to process the println (using a delay adds several (typically five) milliseconds of delay even and the System.out.println only adds a millisecond). This may be caused by some Java behavior I'm not aware of, but still I include it here if it helps.

To try and make the code work, I've taken a look at the code behind the "MethodReader" (since what I'm doing is similar) in order to make sure not to miss something critical but still, it works intermittently. The strangest thing being that when it starts looping forever the method call `w.bytes().readRemaining()` always return the same number (1, 9 or 5).

I've tested multiple version of the library and I cannot get it to work in any of these versions. I'm sorry if this question has already been asked before. I've tried to search but did not really find anybody having a similar issue and certainly not people having the same kind of issues with the current version (4.5.25). Please also forget any English mistakes that might have slipped through since English is not my native language.

The code I'm using for reading is the following (note that this isn't the final code to be used but rather a "proof of concept") :

```java
synchronized(tailer) {
	DocumentContext d = tailer.readingDocument();
	if (!d.isData()) {
		pauser.pause();
	} else {
		pauser.reset();
							
		WireIn w = d.wire();
							
		//reading as long as they are remaining bytes
		while (w.bytes().readRemaining() > 0) {
			//using the same StringBuilder pool as the MethodReader
			StringBuilder sb = WireInternal.acquireStringBuilder(); 	

			ValueIn in = w.readEventName(sb);	

			LockSupport.parkNanos(600); //without this it doesn't work

			//finding a callback based on the event name
			MethodCallback<T> m = callbacks.get(sb.toString()); 
			if (m != null) {
				m.handle(in);
			}
			w.consumePadding();
		}
	}
}
```

**Comment by peter-lawrey** [comment]
Can you create a test which reproduces this issue? This would help diagnose
and fix it.

One thing to note is that ExcerptAppender and ExcerptTailer are single
threaded and cannot be shared between threads without additional thread
safety.
ᐧ

**Comment by AzSeb** [comment]
Hi,
First, I want to thank you for the quick reply and to apologize for my own stupidity since I manage to fix the problem by just thinking on what you said.

I don't think writing a text fixture is necessary, after what you said regarding thread safety, I started adding "synchronized" keywords everywhere and I also stopped using the `WireInternal.acquireStringBuilder();` line. And now, it doesn't even require a `Thread.yield();` or any delay of any sort.

For anyone who might read this and be interested, I changed two major things (both are necessary since it didn't work with either, just with both) in my code: 
* First, I added synchronized keywords everywhere (which did have an effet - I edited my original question to include the synchronized statement I was already using but removed from the original question for the sake of making it shorter) ;
* And second, I changed the way I was reading the data from the `ValueIn in = w.readEventName(sb);`. Since I was generating the MethodCallback (see my code in the op) instances using ASM, I included the logic of reading the value from **the ValueIn instance** in the ASM generated classes but, it was the most stupid thing to do. Since I was actually passing **the instance** of the ValueIn to **another thread** (that you can't see from that code) leaving the ValueIn with unread data thus causing my code looping forever and second, causing the other thread to sometimes read corrupted data from the ValueIn. The reason it sometimes worked was out of pure luck since sometimes, the code (in the other thread) ran quickly enough so that the data in the ValueIn would still be present and not have been cleaned thus letting it work properly.

To recap, the mistake was entirely mine and caused by the multi-threaded code I was running, but after hours of searching for a solution on my own, I took to ask here instead of thinking about what I just explained (derp). Sorry of the inconvenience and again, thank you for the quick answer.

**Comment by peter-lawrey** [comment]
Thank you for your  notes:

I have made WireInternal.acquireStringBuilder(); package local now as this
is could be used internally in unpredicable ways.

ValueIn is highly stateful and not thread safe, so you can't pass it to
another thread.

Thank you for your feedback.

ᐧ

**Comment by dpisklov** [comment]
Closing unless TS disagrees
